pkg/spoke/statusfeedback/reader.go

```diff
@@ -0,0 +1,121 @@
+package statusfeedback
+
+import (
+	"fmt"
+	"reflect"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/client-go/util/jsonpath"
+	workapiv1 "open-cluster-management.io/api/work/v1"
+	"open-cluster-management.io/work/pkg/spoke/statusfeedback/rules"
+)
+
+type StatusReader struct {
+	wellKnownStatus rules.WellKnownStatusRuleResolver
+}
+
+func NewStatusReader() *StatusReader {
+	return &StatusReader{
+		wellKnownStatus: rules.DefaultWellKnownStatusRule(),
+	}
+}
+
+func (s *StatusReader) GetValuesByRule(obj *unstructured.Unstructured, rule workapiv1.FeedbackRule) ([]workapiv1.FeedbackValue, error) {
+	errs := []error{}
+	values := []workapiv1.FeedbackValue{}
+
+	switch rule.Type {
+	case workapiv1.WellKnownStatusType:
+		gvk := obj.GroupVersionKind()
+		paths := s.wellKnownStatus.GetPathsByKind(gvk.GroupKind())
+		if len(paths) == 0 {
+			break
+		}
+
+		for _, path := range paths {
+			value, err := getValueByJsonPath(path.Name, path.Path, obj)
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+			if value == nil {
+				continue
+			}
+			values = append(values, *value)
+		}
+	case workapiv1.JSONPathsType:
+		for _, path := range rule.JsonPaths {
+			// skip if version is specified and the object version does not match
+			if len(path.Version) != 0 && obj.GroupVersionKind().Version != path.Version {
+				continue
+			}
+
+			value, err := getValueByJsonPath(path.Name, path.Path, obj)
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+			if value == nil {
+				continue
+			}
+			values = append(values, *value)
+		}
+	}
+
+	return values, utilerrors.NewAggregate(errs)
+}
+
+func getValueByJsonPath(name, path string, obj *unstructured.Unstructured) (*workapiv1.FeedbackValue, error) {
+	j := jsonpath.New(name).AllowMissingKeys(true)
+	err := j.Parse(fmt.Sprintf("{.status%s}", path))
+	if err != nil {
+		return nil, err
+	}
+
+	results, err := j.FindResults(obj.UnstructuredContent())
+
+	if err != nil {
+		return nil, err
+	}
+
+	if len(results) == 0 || len(results[0]) == 0 {
```

Comment:
what if a `[][]interface{} = [[null]]` result? (iiuc the current implementation will make the controller constantly retrying in failure). i think it should be identical to a `[[]]` b/c mostly the only difference between them is whether the go struct has an `omitempty` in the definition. 

and i can think we should add more tests to pin the behavior to enumerate how different feedback data types will be dealt with, including the failure path. ps we can do the testing in the following ups.